v3/src/transports/http/handler.ts

```diff
@@ -0,0 +1,209 @@
+ *
+ * @param res
+ * @param errorMessage
+ */
+function internalServerError(
+  res: ServerResponse,
+  errorMessage: string = "Internal Server Error",
+): void {
+  res.writeHead(500, errorMessage);
+  res.end()
+}
+
+/**
+ * Take an `http.IncomingMessage` and translate it into a `GraphQLRequest` which
+ * is suitable for consumption by the HTTP transport.  This is a bare-bones
+ * replacement for using a more full-featured package like the popular
+ * [`body-parser`](https://npm.im/body-parser) package.  Of notable absence,
+ * this method does nothing to strictly enforce body length limits, and has
+ * no other error handling.  The `JSON.parse` will of course throw with
+ * malformed input!
+ *
+ * @param req The request from an `http.IncomingMessage` compatible interface.
+ *            (Note that Express' `req` **is** compatible!)
+ *
+ * @throws {SyntaxError}  A malformed request body that does not contain valid
+ *                        JSON structure will cause `JSON.parse` to fail.  This
+ *                        will result in a `SyntaxError` being thrown.
+ *
+ * @remarks
+ *
+ * TODO(AS3) Consider whether this implementation should be used as a getting
+ * started experience that doesn't require an external package.
+ *
+ */
+async function parseRequest(req: IncomingMessage): Promise<GraphQLRequest> {
+  /**
+   * First we'll parse the body, however this doesn't parse objects within
+   * that body, like `variables` and `extensions`, which will need to be further
+   * unwrapped from their JSON encoding to form a processable `GraphQLRequest`.
+   * The blanks in this object will be filled in, as we parse those attributes.
+   * Note that `query` will not be parsed into a `DocumentNode` at this state,
+   * but instead deeper inside GraphQL execution, beyond the transport.
+   */
+  let stringifiedBody: SerializedGraphqlRequest;
+  if (req.method === 'POST') {
+    stringifiedBody = await parsePostRequest(req);
+  } else if (req.method === "GET") {
+    stringifiedBody = await parseGetRequest(req);
+  } else {
+    throw new Error(`Unsupported HTTP method '${req.method}'`)
+  }
+
+  // We'll destructure some parts into `parsedBody`, and save the rest for
+  // further processing.
+  const parsedBody: Partial<GraphQLRequest> = Object.create(null);
+  let unparsedVariables: string | undefined;
+  let unparsedExtensions: string | undefined;
+  ({
+    query: parsedBody.query,
+    operationName: parsedBody.operationName,
+    variables: unparsedVariables,
+    extensions: unparsedExtensions,
+  } = stringifiedBody);
+
+  if (unparsedVariables) {
+    parsedBody.variables = parseJsonInputAsObject(unparsedVariables,
+      "Malformed JSON input for 'variables'");
+  }
+
+  if (unparsedExtensions) {
+    parsedBody.extensions = parseJsonInputAsObject(unparsedExtensions,
+      "Malformed JSON input for 'extensions'");
+  }
+
+  return parsedBody;
+}
+
+export const __testing__ = {
```

Comment:
I really like this style for minimizing the API exposure for testing purposes.

Reply:
Me too! One comment on it: Where I've done it, I find it a bit easier to include the module name in these. So it's like `export const __handler_testing__`, so that if a test imports multiple testpoints, they're easily distinguished.